### ISSUE TITLE
Sourcing an R script using reticulate fails (possibly Win10 specific?)

### ISSUE THREAD

**Issue by erikli** [rstudio/reticulate#584]
Sourcing a simple R script using `reticulate` essentially only runs to the first python command which produces output.

testreticulate.R
```
library(reticulate)

py_run_string("testval=5")
py_run_string("print('Hello World')")
py_run_string("testval=6")
```

R session:
```
> source("testreticulate.R")
Error in sys$stdout$flush() : attempt to apply non-function
> py$testval
[1] 5
```

Tracked the error down to https://github.com/rstudio/reticulate/issues/284 and 3dc08ec70c4078189bf584fc1098575c9681495c -- on Windows 10 and R 3.6.1 at least it seems that `sys$stdout` and `sys$stderr` are null so calling `sys$stdout$flush()` tries to call `flush()` on `NULL`

```
> sys=import("sys")
> sys$stdout
NULL
> sys$stderr
NULL
> sys$stderra
Error in py_get_attr_impl(x, name, silent) : 
  AttributeError: module 'sys' has no attribute 'stderra'
```

I currently fixed this for myself in https://github.com/erikli/reticulate/commit/8a7b8a57ff3ba5b985e92eb40a823eb07c899133 by just commenting out the flushes (which are now in python.R), but I think a more cross-platform way to deal might be to call `py_run_string("sys.stdout.flush()")` or similar? Interestingly, on my computer at least, I still get output `Hello World` even without the explicit flush.

```

Type 'q()' to quit R.

> library(devtools)
Loading required package: usethis
> devtools::install_github("erikli/reticulate")
Downloading GitHub repo erikli/reticulate@master
Your system is ready to build packages!
These packages have more recent versions available.
Which would you like to update?

1: All                         
2: CRAN packages only          
3: None                        
4: Rcpp (1.0.1 -> 1.0.2) [CRAN]

Enter one or more numbers, or an empty line to skip updates:
3
  
  
  
   checking for file 'C:\Users\Erik\AppData\Local\Temp\RtmpCYipB2\remotes34103e824328\erikli-reticulate-8a7b8a5/DESCRIPTION' ...
  
√  checking for file 'C:\Users\Erik\AppData\Local\Temp\RtmpCYipB2\remotes34103e824328\erikli-reticulate-8a7b8a5/DESCRIPTION'

  
  
  
-  preparing 'reticulate': (601ms)
   checking DESCRIPTION meta-information ...
  
   checking DESCRIPTION meta-information ... 
  
√  checking DESCRIPTION meta-information
-  cleaning src

  
  
  
-  checking for LF line-endings in source and make files and shell scripts

  
  
  
-  checking for empty or unneeded directories

  
  
  
-  building 'reticulate_1.13.0-9000.tar.gz'

  
   

Installing package into ‘C:/Users/Erik/Documents/R/win-library/3.6’
(as ‘lib’ is unspecified)
* installing *source* package 'reticulate' ...
** using staged installation
** libs

*** arch - i386
C:/Rtools/mingw_32/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c RcppExports.cpp -o RcppExports.o
C:/Rtools/mingw_32/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c event_loop.cpp -o event_loop.o
C:/Rtools/mingw_32/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c libpython.cpp -o libpython.o
C:/Rtools/mingw_32/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c output.cpp -o output.o
C:/Rtools/mingw_32/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c python.cpp -o python.o
C:/Rtools/mingw_32/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c readline.cpp -o readline.o
C:/Rtools/mingw_32/bin/g++ -shared -s -static-libgcc -o reticulate.dll tmp.def RcppExports.o event_loop.o libpython.o output.o python.o readline.o -LC:/PROGRA~1/R/R-36~1.1/bin/i386 -lR
installing to C:/Users/Erik/Documents/R/win-library/3.6/00LOCK-reticulate/00new/reticulate/libs/i386

*** arch - x64
C:/Rtools/mingw_64/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c RcppExports.cpp -o RcppExports.o
C:/Rtools/mingw_64/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c event_loop.cpp -o event_loop.o
C:/Rtools/mingw_64/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c libpython.cpp -o libpython.o
C:/Rtools/mingw_64/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c output.cpp -o output.o
C:/Rtools/mingw_64/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c python.cpp -o python.o
C:/Rtools/mingw_64/bin/g++  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG  -I"C:/Users/Erik/Documents/R/win-library/3.6/Rcpp/include"        -O2 -Wall  -mtune=generic -c readline.cpp -o readline.o
C:/Rtools/mingw_64/bin/g++ -shared -s -static-libgcc -o reticulate.dll tmp.def RcppExports.o event_loop.o libpython.o output.o python.o readline.o -LC:/PROGRA~1/R/R-36~1.1/bin/x64 -lR
installing to C:/Users/Erik/Documents/R/win-library/3.6/00LOCK-reticulate/00new/reticulate/libs/x64
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
  converting help for package 'reticulate'
    finding HTML links ... done
    array_reshape                           html  
    as.character.python.builtin.bytes       html  
    conda-tools                             html  
    dict                                    html  
    eng_python                              html  
    import                                  html  
    iterate                                 html  
    np_array                                html  
    py                                      html  
    py_available                            html  
    py_call                                 html  
    py_capture_output                       html  
    py_config                               html  
    py_config_error_message                 html  
    py_del_item                             html  
    py_discover_config                      html  
    py_func                                 html  
    py_function_custom_scaffold             html  
    py_function_wrapper                     html  
    py_get_attr                             html  
    py_get_item                             html  
    py_has_attr                             html  
    py_help                                 html  
    py_id                                   html  
    py_install                              html  
    py_is_null_xptr                         html  
    py_iterator                             html  
    py_last_error                           html  
    py_len                                  html  
    py_list_attributes                      html  
    py_main_thread_func                     html  
    py_module_available                     html  
    py_run                                  html  
    py_save_object                          html  
    py_set_attr                             html  
    py_set_item                             html  
    py_set_seed                             html  
    py_str                                  html  
    py_suppress_warnings                    html  
    py_to_r_wrapper                         html  
    py_unicode                              html  
    py_versions_windows                     html  
    r-py-conversion                         html  
    register_class_filter                   html  
    register_help_topics                    html  
    register_module_help_handler            html  
    register_suppress_warnings_handler      html  
    repl_python                             html  
    reticulate                              html  
    source_python                           html  
    tuple                                   html  
    use_python                              html  
    virtualenv-tools                        html  
    with-as-operator                        html  
    with.python.builtin.object              html  
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
*** arch - i386
*** arch - x64
** testing if installed package can be loaded from final location
*** arch - i386
*** arch - x64
** testing if installed package keeps a record of temporary installation path
* DONE (reticulate)
> setwd("../Desktop")
> source("testreticulate.R")
Hello World
> py$testval
[1] 6
```

**Comment by erikli** [comment]
Possibly related to https://stackoverflow.com/questions/57357001/reticulate-1-13-error-in-sysstdoutflush-attempt-to-apply-non-function in which case it is specific to Rgui and doesn't effect RStudio

**Comment by BurdenBear** [comment]
Same error in windows 10 or window server R2016,   R studio  1.1.463  , R 3.6.1 32bit.

**Comment by skeydan** [comment]
Hi,

can people who are affected by this (also applying to #662) please paste the output of 

```
reticulate::py_config()
```

as well as test if setting the environment variable (e.g. in .Renviron)

```
RETICULATE_REMAP_OUTPUT_STREAMS = 1
```
helps?

Given that the condition for remapping is Windows AND Python 2

https://github.com/rstudio/reticulate/blob/9d973bfc9db12e983e3d64ed0b38cd9b04c10366/R/output.R#L6-L18

the only explanation that comes to mind is that on some Windows systems, this also happens with Python 3?

**Comment by TylerGrantSmith** [comment]
I am getting the same error even with `RETICULATE_REMAP_OUTPUT_STREAMS = 1` in `.Renviron`.

```
python:         C:/Users/e014307/AppData/Local/Continuum/anaconda3/envs/torch/python.exe
libpython:      C:/Users/e014307/AppData/Local/Continuum/anaconda3/envs/torch/python37.dll
pythonhome:     C:/Users/e014307/AppData/Local/Continuum/anaconda3/envs/torch
version:        3.7.5 (default, Oct 31 2019, 15:18:51) [MSC v.1916 64 bit (AMD64)]
Architecture:   64bit
numpy:          C:/Users/e014307/AppData/Local/Continuum/anaconda3/envs/torch/lib/site-packages/numpy
numpy_version:  1.17.4
sys:            [builtin module]

python versions found: 
 C:/Users/e014307/AppData/Local/Continuum/anaconda3/envs/torch/python.exe
 C:/Users/e014307/AppData/Local/Continuum/anaconda3/envs/r-reticulate/python.exe
 C:/Users/e014307/AppData/Local/Continuum/anaconda3/python.exe
 C:/Users/e014307/AppData/Local/Programs/Python/Python38/python.exe
```

**Comment by skeydan** [comment]
Thanks! Should be addressed in

https://github.com/rstudio/reticulate/pull/673

**Comment by skeydan** [comment]
Hi,

this should be fixed in the current dev version of `reticulate` (master branch on github).

I'm closing the issue, please feel free to reopen if you still run into problems.

**Comment by JohnGavin** [comment]
Just in case it helps any else 
RETICULATE_REMAP_OUTPUT_STREAMS = 1
fixed the same issue for me on python 3.7 on a mac.

~/.Renviron
RETICULATE_REMAP_OUTPUT_STREAMS = 1

>>> for i in range(5):
...     print(str(i))
... 
0
1
2
3
4
>>>

Without setting RETICULATE_REMAP_OUTPUT_STREAMS, 
it is necessary to use flush with every print, 
else nothing is printed (until I quit python back to R).

>>> for i in range(5):
...     print(str(i), flush=True)
... 
0
1
2
3
4
>>>

> py_config()
python:         /Users/jbg/Library/r-miniconda/envs/r-reticulate/bin/python
libpython:      /Users/jbg/Library/r-miniconda/envs/r-reticulate/lib/libpython3.7m.dylib
pythonhome:     /Users/jbg/Library/r-miniconda/envs/r-reticulate:/Users/jbg/Library/r-miniconda/envs/r-reticulate
version:        3.7.7 (default, Mar 23 2020, 17:31:31)  [Clang 4.0.1 (tags/RELEASE_401/final)]
numpy:          /Users/jbg/Library/r-miniconda/envs/r-reticulate/lib/python3.7/site-packages/numpy
numpy_version:  1.18.5
NOTE: Python version was forced by RETICULATE_PYTHON

> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] here_0.1             testthat_2.3.2       forcats_0.5.0       
 [4] stringr_1.4.0        dplyr_1.0.0          purrr_0.3.4         
 [7] readr_1.3.1          tidyr_1.1.0          tibble_3.0.1        
[10] ggplot2_3.3.2        tidyverse_1.3.0      reticulate_1.16-9000

loaded via a namespace (and not attached):
 [1] tidyselect_1.1.0 xfun_0.15        splines_4.0.2    haven_2.3.1     
 [5] lattice_0.20-41  colorspace_1.4-1 vctrs_0.3.1      generics_0.0.2  
 [9] htmltools_0.5.0  mgcv_1.8-31      yaml_2.2.1       blob_1.2.1      
[13] rlang_0.4.6      pillar_1.4.4     withr_2.2.0      glue_1.4.1      
[17] DBI_1.1.0        dbplyr_1.4.4     modelr_0.1.8     readxl_1.3.1    
[21] lifecycle_0.2.0  munsell_0.5.0    gtable_0.3.0     cellranger_1.1.0
[25] rvest_0.3.5      evaluate_0.14    labeling_0.3     knitr_1.29      
[29] fansi_0.4.1      broom_0.5.6      Rcpp_1.0.4.6     backports_1.1.8 
[33] scales_1.1.1     jsonlite_1.7.0   farver_2.0.3     fs_1.4.1        
[37] hms_0.5.3        png_0.1-7        packrat_0.5.0    digest_0.6.25   
[41] stringi_1.4.6    grid_4.0.2       rprojroot_1.3-2  cli_2.0.2       
[45] tools_4.0.2      magrittr_1.5     crayon_1.3.4     pkgconfig_2.0.3 
[49] ellipsis_0.3.1   Matrix_1.2-18    xml2_1.3.2       reprex_0.3.0    
[53] lubridate_1.7.9  assertthat_0.2.1 rmarkdown_2.3    httr_1.4.1      
[57] rstudioapi_0.11  R6_2.4.1         nlme_3.1-148     compiler_4.0.2
